ta/crypto_perf/ta_crypto_perf.c

```diff
@@ -465,8 +467,482 @@ TEE_Result cmd_hash_prepare_op(uint32_t param_types, TEE_Param params[4])
+
+	cmd_clean_res();
+	res = TEE_AllocateOperation(&crypto_op, algo, mode, width_bits);
+	CHECK(res, "TEE_AllocateOperation()", goto out;);
+
+	res = TEE_SetOperationKey(crypto_op, crypto_obj);
+	CHECK(res, "TEE_SetOperationKey()", goto out;);
+
+	if (mode == MODE_DECRYPT) {
+		res = TEE_AllocateOperation(&crypto_op_enc_sign, algo,
+					    MODE_ENCRYPT, width_bits);
+		CHECK(res, "TEE_AllocateOperation()", goto out;);
+
+		res = TEE_SetOperationKey(crypto_op_enc_sign, crypto_obj);
+		CHECK(res, "TEE_SetOperationKey()", goto out;);
+	} else if (mode == MODE_VERIFY) {
+		res = TEE_AllocateOperation(&crypto_op_enc_sign, algo,
+					    MODE_SIGN, width_bits);
+		CHECK(res, "TEE_AllocateOperation()", goto out;);
+
+		res = TEE_SetOperationKey(crypto_op_enc_sign, crypto_obj);
+		CHECK(res, "TEE_SetOperationKey()", goto out;);
+	}
+
+out:
+	TEE_Free(attrs);
+
+	return res;
+}
+
+TEE_Result cmd_asym_prepare_hash(uint32_t param_types,
+				 TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_OperationHandle hash_op = NULL;
+	uint32_t hash_algo = 0;
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_MEMREF_INPUT,
+						   TEE_PARAM_TYPE_MEMREF_INOUT,
+						   TEE_PARAM_TYPE_NONE);
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (params[0].value.a == ECDSA)
+		hash_algo = TEE_ALG_SHA1;
+	else
+		hash_algo = TEE_ALG_HASH_ALGO(TEE_ALG_GET_DIGEST_HASH(algo));
+
+	res = TEE_AllocateOperation(&hash_op, hash_algo, TEE_MODE_DIGEST, 0);
+	CHECK(res, "TEE_AllocateOperation()", return res;);
+
+	res = TEE_DigestDoFinal(hash_op, params[1].memref.buffer,
+				params[1].memref.size, params[2].memref.buffer,
+				&params[2].memref.size);
+	TEE_FreeOperation(hash_op);
+	CHECK(res, "TEE_DigestDoFinal()", return res;);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result cmd_asym_prepare_enc_sign(uint32_t param_types,
+				TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_Attribute *attrs = NULL;
+	uint32_t attr_count = 0;
+	uint32_t mode = 0;
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+						   TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						   TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_MEMREF_INPUT);
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mode = params[2].value.a;
+	res = unpack_attrs(params[3].memref.buffer, params[3].memref.size,
+			   &attrs, &attr_count);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	if (mode == MODE_DECRYPT)
+		res = TEE_AsymmetricEncrypt(crypto_op_enc_sign, NULL, 0,
+					    params[0].memref.buffer, params[0].memref.size,
+				            params[1].memref.buffer, &params[1].memref.size);
+	else
+		res = TEE_AsymmetricSignDigest(crypto_op_enc_sign, attrs,
+					       attr_count,
+					       params[0].memref.buffer,
+					       params[0].memref.size,
+					       params[1].memref.buffer,
+					       &params[1].memref.size);
+
+	TEE_Free(attrs);
```

Comment:
using global variables for attribute could simplify the cleanup code.
here we have to call TEE_Free before testing return value, duplicating if statements.
with attrs in global, all memory cleanup could be done in the close session as already done for objects and operations.

Reply:
I think the way it is is fine. This is temp buffer used to prepare the crypto operation, it's fine being allocated and freed here.